### PR TITLE
fix: windows auto-update

### LINKED ISSUE #1570: Auto-update broken on Windows
This issue is directly related to issue #1514 .  Not sure why this was closed before verifying that it worked, because, when I initiated the auto update to ipfs-desktop 12.0, the exact same issue occurred once again. Program was uninstalled, not re-installed, and the installation files were located in the %USERPROFILE%\Appdata\local\ipfs-desktop-updater. I did notice that there were 2 executable files in the directory though (I don't recall if there were last time): one named "updater.exe" and one that was the actual install file for the program (IPFS-Desktop-Setup-0.12.0.exe). Interestingly, while both were the same size, (83,195 kb), I noticed that the hash values were different, so they are not the same file? I don't know. What I do know is that the process is still broken and the potential cause of the issue which I mentioned in the 2nd of the 2 potential causes I listed remains unchanged (update-info.json "isAdminRightsRequired" value is still euqal to false, and as I stated the first time, you need admin permissions to install an application to the program files directory, which is where ipfs desktop is installed when you install it for all users). Maybe I am an outlier in this mindset, but I am of the belief that issues should be confirmed resolved before closing them.Auto update not working on Windows when installed globally for all users

 - OS: Windows 10
 - Version of IPFS Desktop [e.g. 0.12.0]

**Describe the bug**
I have noticed that, the last two times that IPFS-desktop has updated using the internal update functionality (when the app says a new version is available and to click button to close IPFS-desktop and install the updated version), the old version of ipfs-desktop successfully uninstalls the old version, but does not install the new version **correctly.**  I should note that I install IPFS-desktop so that it available for all users, not just the currently running user, which I think is the main reason the auto-update might be failing.  There are a few reasons that I can think of that might be the cause:

1. In my case, when I installed IPFS, I installed it so that it would be available for all users.  As you may know, this is not the default configuration for the installer (the default being to install IPFS-desktop only for the current user), so what **MAY** be happening is that IPFS-desktop is uninstalling the old version located in Program files (the folder to install app for all users), but then installing the new version in app data (only install for current user).  If this turns out to be the case, while technically still successfully installing the application (which is why I emphasized "correctly"), my IPFS-desktop Desktop icon can no longer find its source executable, and I just end up downloading the installer and reinstalling the application manually anyway, defeating the auto update functionality entirely.  This is unlikely the cause.
2. The process initiating that auto-update functionality may not have the elevated permissions that it needs to write to the Program files directory.  If the installer takes into account the install settings from the previous installation (in other words, installs the new version for all users if the previous install was configured to be installed for all users), then the reason that it might be failing is because the process may lack the elevated permissions needed to write to the Program Files directory.  I am confident that this is the cause of the failed installation, as, when I navigated to %USERNAME%\appdata\local\ipfs-desktop-updater\pending and opened the update-info.json file, I saw the following:

image

**To Reproduce**
1. Install ipfs-desktop version 0.11.2 or 0.11.1 and configure the installation so that the application is available to all users
2. Launch the application
3. Wait for the auto-update functionality dialog box to appear and initiate the process

**Expected behavior**
IPFS auto-update functionality would automatically re-install the application successfully with the same configurations as the previous install

[error.log](https://github.com/ipfs-shipyard/ipfs-desktop/files/4646202/error.log)

### PR DESCRIPTION
**Pull request by lidel**
This PR aims to close #1570
Context: https://github.com/ipfs-shipyard/ipfs-desktop/issues/1570#issuecomment-705081315

## WIP

- [x] revert #1556 to reduce noise
- [x] figure out local setup for testing autoupdate on windows
  - We won't know for sure, unless we  test in production, but we don't need to use this app
  - I created [sandbox electron app](https://github.com/lidel/electron-updater-altschmerz), which uses electron-build config similar to one in ipfs-desktop and basically the same CI setup
  - Next step is to copy autoupdate code from ipfs-desktop to that app, and see how it behaves on Windows.
- [x] see if the  latest `electron-builder` fixed the original problem upstream
  - no, still broken https://github.com/ipfs-shipyard/ipfs-desktop/pull/1679#issuecomment-705569228
- [x] if still broken: fix windows autoupdate 
  - fix: https://github.com/ipfs-shipyard/ipfs-desktop/pull/1679#issuecomment-705630973
- [x] confirm it works when installed for single user
- [x] confirm it works when installed for every user

cc @jessicaschilling @rafaelramalho19

### CODE REVIEW

**Comment by jessicaschilling** [comment]
@lidel - is this related to https://github.com/ipfs-shipyard/ipfs-desktop/issues/1677?

**Comment by lidel** [comment]
@jessicaschilling unclear. that error could be a side effect of autoupdate issue, or a duplicate of #1676 – asked for more information.

**Comment by lidel** [comment]
Closing the shop for today, but quick update:

I looked into setting up local update server, but it effectively becomes _Rube Goldberg machine_ and I don't believe it is a reliable way of testing and debugging this. To be sure we test the right thing, we literally need to test _in production_.

For that reason I created a [test app](https://github.com/lidel/electron-updater-altschmerz) with CI and electron config similar to Desktop. 

I'll copy relevant autoupdate code from Desktop tomorrow, so it is identical, and then we have setup to run some end-to-end tests.

**Comment by lidel** [comment]
I tested the simplified autoupdate provided by `electron-updater` on windows and it works fine when app is installed for current user, but fails when "all users" option is selected during install:

> Capture-user-picker


This means original bug is still present in `electron-updater`. 
I'll now check if a viable fix exists, if not, we have to either disable "all users" option or detect it and disable automatic install, asking user for manual update instead"

**Comment by lidel** [comment]
Disclaimer: I am not a Windows person, so this may be obvious stuff, but took me a while to understand all the moving pieces. Having the [altschmerz test app](https://github.com/lidel/electron-updater-altschmerz) helped a lot.

I believe the crux of the problem was that we use custom NSIS settings, namely:

```yaml
nsis:
  oneClick: false
  warningsAsErrors: false
  perMachine: false
  allowToChangeInstallationDirectory: true
```

By default we install per user (`perMachine: false`), but we also give user an option to change that via the screen from my previous comment (`oneClick: false`).

We don't want to require admin all the time, because users should be able to install IPFS on machines where they don't have Admin (schools, libraries). That is why `isAdminRightsRequired` is `false` in the update manifest (as described in https://github.com/ipfs-shipyard/ipfs-desktop/issues/1570#issue-652502677). 

`electron-builder` has a bunch of default macros for NSIS, and some higher level flags that control some Windows-specific behaviours, for example, [`electron-builder/**/nsis/NsisTarget.ts#L438-L444`](https://github.com/electron-userland/electron-builder/blob/v22.8.1/packages/app-builder-lib/src/targets/nsis/NsisTarget.ts#L438-L444) sets:

```js
    if (options.perMachine === true) {
      defines.INSTALL_MODE_PER_ALL_USERS = null
    }

    if (!oneClick || options.perMachine === true) {
      defines.INSTALL_MODE_PER_ALL_USERS_REQUIRED = null
    }
```

`INSTALL_MODE_PER_ALL_USERS` and `INSTALL_MODE_PER_ALL_USERS_REQUIRED` are dynamic flags that control various things, one of them is deciding if installer will ask windows for elevated privileges before doing its thing.

In our case, `INSTALL_MODE_PER_ALL_USERS` is ON and `INSTALL_MODE_PER_ALL_USERS_REQUIRED` is OFF, and `UAC_RunElevated` is only used in `_REQUIRED` scenario – see [`electron-builder/**/nsis/multiUser.nsh#L52-L62`](https://github.com/electron-userland/electron-builder/blob/b8d6905582b28a9b36f11308f9a72ebc32dc2c86/packages/app-builder-lib/templates/nsis/multiUser.nsh#L52-L62)


**TLDR** the fix here is to 
- (a) fix upstream 
- (b) extend `assets/build/nsis.nsh` with `customInit` script that requests `UAC_RunElevated` when `${if} $installMode == "all"` (independent of `INSTALL_MODE_PER_ALL_USERS_REQUIRED`)

I'll prepare (b) first, as (a) will take long time before the fix lands upstream, and I still don't trust there won't be any regressions, as we had so many in the past, unfortunately.

**Comment by lidel** [comment]
Implemented the fix in  0bf120d – ok to merge and ship patch release?

**Comment by rafaelramalho19** [comment]
Tested and looks good, but we can only really find out on the next version... so 🚢 it !

**Review by rafaelramalho19 — APPROVED** [review]

**Comment by witcher112** [comment]
Hi guys!

I encountered this PR while resolving some of the UAC issues with electron-builder auto updater.

Just wanted to let you know that there might be an alternative solution - modifying the access rights after the installation, to grant full access permissions to the non-admin users (to be honest not my idea, I just observed what Steam does) so **no administrative rights are required for the updater**.

Here's the snippet for custom NSIS include:
```
!macro customInstall
  AccessControl::GrantOnFile "$INSTDIR" "(S-1-5-32-545)" "FullAccess"
  pop $0
  AccessControl::GrantOnFile "$INSTDIR" "(S-1-5-18)" "FullAccess"
  pop $0
!macroend
```

_`S-1-5-32-545` is a SID for Users group_
_`S-1-5-18` is a SID for SYSTEM group_

Note that you need to link [AccessControl](https://nsis.sourceforge.io/AccessControl_plug-in) plugin to NSIS either through `!addplugindir` or by putting it into `build/x86-unicode` ([source](https://www.electron.build/configuration/nsis#custom-nsis-script)).

Cheers and good luck with your project!

**Comment by lidel** [comment]
@witcher112 thank you, useful to know there is an alternative approach in case current one stops working.

Maybe I misunderstood the meaning of `GrantOnFile` (or just lacking Windows background), but.. isn't granting write access to non-admin users effectively a decrease in security? 

IIUC one user could overwrite binary that other users run.
